### PR TITLE
Lua Userdata

### PR DESCRIPTION
**Pull request by mynameundefined**
Adds a new Class "LuaUserdata", that can be used to create objects with a C# metatable.
Makes LuaLibrary use LuaUserdata

### CODE REVIEW

File: Lua14/Patches/GetMemberPatch.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
+﻿using HarmonyLib;
+using NLua;
+using System.Reflection;
+using System.Reflection.Emit;
+
+namespace Lua14.Patches;
+
+/// <summary>
+/// fixes "instance method ... requires a non null target object"
+/// </summary>
```
Comment (Subversionary) [inline Lua14/Patches/GetMemberPatch.cs:10]: Explain why this fix is required

File: Lua14/Lua/LuaExtensions.cs [hunk 2]
```diff
@@ -1,11 +1,18 @@
...
 
 namespace Lua14.Lua;
 
 public static class LuaExtensions
 {
+    private static readonly FieldInfo F_Translator = typeof(NLua.Lua)
+        .GetField("_translator", BindingFlags.NonPublic | BindingFlags.Instance) ?? throw new Exception("Field _translator doesnt exist on NLua.Lua");
+    private static readonly MethodInfo M_GetUserData = typeof(ObjectTranslator)
+        .GetMethod("GetUserData", BindingFlags.NonPublic | BindingFlags.Instance) ?? throw new Exception("Method GetUserData doesnt exist on ObjectTranslator");
+
```
Comment (Subversionary) [inline Lua14/Lua/LuaExtensions.cs:12]: Why not AccessTools?

File: Lua14/Patches/GetMemberPatch.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                continue;
+            }
+
+            codes[index + 1] = new CodeInstruction(OpCodes.Ldarg_S, 3); // loads object
+            codes[index + 2] = new CodeInstruction(OpCodes.Ldarg_2); // loads ProxyType
+            codes[index + 3] = new CodeInstruction(OpCodes.Ldloc_1); // loads MemberInfo
+            codes[index + 4] = new CodeInstruction(OpCodes.Newobj, TargetWrapperCtor);
+
+            codes.Insert(index + 4, new CodeInstruction(OpCodes.Castclass, typeof(MethodBase))); // pops MemberInfo and loads MethodBase
+        }
```
Comment (Subversionary) [inline Lua14/Patches/GetMemberPatch.cs:51]: Won't this skip only the instruction preceeding the Ldarg_2, Ldarg_s, Ldarg_S, Newobj, causing those operations to be either orphaned or loaded where they dont belong?

File: Lua14/EntryPoint.cs [hunk 4]
```diff
@@ -10,6 +11,8 @@ public sealed class EntryPoint : GameShared
 {
     public override void PreInit()
     {
+        Harmony.DEBUG = true;
+
```
Comment (Subversionary) [inline Lua14/EntryPoint.cs:15]: While this is not conflicting with Marsey's DebugAllowed flag - this should be either tied to it or logged separately with the Harmony-provided FileLog class.

File: Lua14/Patches/GetMemberPatch.cs (hunk 3 above)
Comment (mynameundefined) [reply Lua14/Patches/GetMemberPatch.cs:51]: It doesn't skip the instruction, instead it finds the needed instruction that loads the translator field by its following instructions and applies patches to them.

Comment (Subversionary) [reply Lua14/Patches/GetMemberPatch.cs:51]: Seems like you'r ereplacing them and then insertinting a new cast codeinstruction, fair.

**Pull request merged**